### ISSUE TITLE
`Button::on_press` should take a closure or there should be an alternative that does that

### ISSUE THREAD

**Issue by zroug** [hecrj/iced#312]
This hits me now and then but most often when writing generic element functions:

```rust
fn cool_element<'a, TMessage: 'static + Clone>(
    state: &'a mut State,
    on_event: impl Fn(u32) -> TMessage + 'a
) -> Element<'a, TMessage> {
    Button::new(
        &mut state.button_state,
        Text::new("Press me!"),
    ).on_press(on_event(42)).into()
}
```

`on_event` is executed on every call to `widget` but the caller of such a function expects that `on_event` gets called only when the event actually happens. Because of that he might do something relatively expensive, that is appropriate after a click but not on every render.

**Comment by ethanpailes** [comment]
@zroug, my understanding is that you are supposed to perform your intensive computations in the `update` routine. What's an example of a computation that must be performed in the view routine and can't be described by an event variant?

**Comment by zroug** [comment]
Hi, sorry that I didn't respond. I'm currently not using Iced anymore, because I'm working on a different project now. That happened shortly after I opened the issue. But my latest experiments showed that my proposed change would not be sufficient for what I wanted to do because of the `'static` message lifetime. I'm closing this issue because I'm not sure if I actually want something changed. Sorry for inconvenience I caused with this issue.
